src/utils/unique_stack.rs

```diff
@@ -0,0 +1,221 @@
+//! A stack that only allows unique elements.
+
+use std::hash::Hash;
+
+use crate::HashMap;
+
+#[derive(Clone)]
+struct El<T> {
+    value: T,
+    below: Option<usize>,
+    above: Option<usize>,
+}
+
+/// Stack with unique elements.
+///
+/// A first in first out stack. If pushing an element that is already
+/// in the stack, the operation moves that element to the top of the stack.
+#[derive(Clone)]
+pub(crate) struct UniqueStack<T> {
+    elements: Vec<Option<El<T>>>,
+    indices: HashMap<T, usize>,
+    top: Option<usize>,
+    free_indices: Vec<usize>,
+}
+
+impl<T: Eq + Hash + Clone> UniqueStack<T> {
+    /// Create a new empty UniqueStack.
+    pub fn new() -> Self {
+        UniqueStack {
+            elements: Vec::new(),
+            indices: HashMap::default(),
+            top: None,
+            free_indices: Vec::new(),
+        }
+    }
+
+    /// Push an element onto the stack, moving it to the top if already present.
+    ///
+    /// Return true if the element was added, false if it was already present.
+    pub fn push(&mut self, value: T) -> bool {
+        if let Some(&index) = self.indices.get(&value) {
+            self.rewire_top(index);
+            false
+        } else {
+            // New element, add it to the top
+            let new_index = self.get_free_index();
+            let el = El {
+                value: value.clone(),
+                below: None,
+                above: None,
+            };
+            self.elements[new_index] = Some(el);
+            self.indices.insert(value, new_index);
+            self.rewire_top(new_index);
+            true
+        }
+    }
+
+    /// Remove and return the top element from the stack.
+    ///
+    /// Return None if the stack is empty.
+    pub fn pop(&mut self) -> Option<T> {
+        let top_index = self.top?;
+        let el = self.elements[top_index].take().unwrap();
+        self.indices.remove(&el.value);
+        self.top = el.below;
```

Comment:
Update the new top's `above`?

Reply:
Thank you for the careful reading, that would have been hard to find later on...